File: modules/react-native-md-list/android/src/main/java/com/mdlist/md/MdLayout.kt
package com.mdlist.md

import android.text.Layout
import android.text.SpannableStringBuilder
import android.text.Spanned
import android.text.StaticLayout
import android.text.TextPaint
import android.text.style.BackgroundColorSpan
import android.text.style.ForegroundColorSpan
import android.text.style.RelativeSizeSpan
import android.text.style.StrikethroughSpan
import android.text.style.StyleSpan
import android.text.style.TypefaceSpan
import android.graphics.Typeface
import android.util.LruCache
import kotlin.math.ceil
import kotlin.math.max
import kotlin.math.min

/** Marker span carrying the URL; hit tested manually so we avoid ClickableSpan/TextView. */
class MdLinkSpan(@JvmField val url: String)

sealed class RowLayout {
    abstract val height: Int
}

class TextRowLayout(
    @JvmField val text: Spanned,
    @JvmField val layout: StaticLayout,
    @JvmField val textX: Float,
    @JvmField val textY: Float,
    @JvmField val marker: StaticLayout?,
    @JvmField val markerX: Float,
    @JvmField val quoteDepth: Int,
    @JvmField val quoteX: Float,
    @JvmField val checked: Int,
    override val height: Int,
) : RowLayout()

class CodeRowLayout(
    @JvmField val language: String,
    @JvmField val code: String,
    @JvmField val text: Spanned,
    @JvmField val layout: StaticLayout,
    @JvmField val headerHeight: Int,
    @JvmField val contentWidth: Int,
    @JvmField val padding: Int,
    @JvmField val left: Int,
    @JvmField val boxWidth: Int,
    @JvmField val topMargin: Int,
    override val height: Int,
) : RowLayout()

class TableRowLayout(
    @JvmField val cells: Array<Array<StaticLayout>>,
    @JvmField val colW: IntArray,
    @JvmField val rowH: IntArray,
    @JvmField val contentWidth: Int,
    @JvmField val cellPad: Int,
    @JvmField val left: Int,
    @JvmField val topMargin: Int,
    override val height: Int,
) : RowLayout()

class DividerRowLayout(@JvmField val inset: Float, override val height: Int) : RowLayout()

class BubbleRowLayout(
    @JvmField val text: Spanned,
    @JvmField val layout: StaticLayout,
    @JvmField val bubbleLeft: Float,
    @JvmField val bubbleTop: Float,
    @JvmField val bubbleWidth: Float,
    @JvmField val bubbleHeight: Float,
    @JvmField val pad: Float,
    override val height: Int,
) : RowLayout()

/**
 * Builds and caches the text layout of every row.
 *
 * Text measurement is *the* cost in a markdown list. Doing it here means:
 *  - it happens on a worker thread (StaticLayout is safe to build off the UI thread),
 *  - it happens once per (row content, width) and is then reused by every bind,
 *  - onMeasure of a recycled row is O(1): the height is already known.
 */
class MdLayoutEngine(@JvmField val theme: MdTheme) {

    private val cache = object : LruCache<String, RowLayout>(512) {}

    @JvmField val hPad: Float = theme.dp(16f)
    private val quoteInset = theme.dp(14f)
    private val listIndent = theme.dp(18f)
    private val markerGutter = theme.dp(20f)

    fun clear() = cache.evictAll()

    fun cached(row: MdRow, width: Int): RowLayout? = cache.get(key(row, width))

    fun layoutFor(row: MdRow, width: Int): RowLayout {
        val k = key(row, width)
        cache.get(k)?.let { return it }
        val built = build(row, width)
        cache.put(k, built)
        return built
    }

    private fun key(row: MdRow, width: Int) = "${row.layoutKey}|$width|${theme.key}"

    private fun build(row: MdRow, width: Int): RowLayout = when (row.type) {
        RowType.BUBBLE -> buildBubble(row, width)
        RowType.CODE -> buildCode(row, width)
        RowType.TABLE -> buildTable(row, width)
        RowType.DIVIDER -> DividerRowLayout(hPad, (theme.dp(25f)).toInt())
        else -> buildText(row, width)
    }

    // ------------------------------------------------------------------ text

    private fun buildText(row: MdRow, width: Int): RowLayout {
        val isQuote = row.quoteDepth > 0
        val paint: TextPaint = when {
            row.type == RowType.HEADING -> theme.headingPaint(row.headingLevel)
            isQuote -> theme.quote
            else -> theme.body
        }

        val quoteX = hPad
        val quotePad = row.quoteDepth * quoteInset
        var left = hPad + quotePad
        var marker: StaticLayout? = null
        var markerX = 0f

        if (row.type == RowType.LIST_ITEM) {
            left += row.listDepth * listIndent
            markerX = left
            if (row.checked < 0) {
                val m = SpannableStringBuilder(row.listMarker ?: "•")
                marker = staticLayout(m, markerPaint(paint), ceil(markerGutter).toInt(), Layout.Alignment.ALIGN_NORMAL)
            }
            left += markerGutter
        }

        val avail = max(1, (width - left - hPad).toInt())
        val spanned = spannable(row.inline, paint)
        val layout = staticLayout(spanned, paint, avail, Layout.Alignment.ALIGN_NORMAL)

        val topPad = topPadFor(row)
        val bottomPad = bottomPadFor(row)
        val height = (topPad + layout.height + bottomPad).toInt()

        return TextRowLayout(
            text = spanned,
            layout = layout,
            textX = left,
            textY = topPad,
            marker = marker,
            markerX = markerX,
            quoteDepth = row.quoteDepth,
            quoteX = quoteX,
            checked = row.checked,
            height = height,
        )
    }

    private fun markerPaint(base: TextPaint) = base

    private fun topPadFor(row: MdRow): Float {
        val extra = if (row.isFirstInMessage) theme.dp(10f) else 0f
        return extra + when (row.type) {
            RowType.HEADING -> when (row.headingLevel) {
                1 -> theme.dp(18f); 2 -> theme.dp(16f); 3 -> theme.dp(14f); else -> theme.dp(12f)
            }
            RowType.LIST_ITEM -> theme.dp(3f)
            else -> theme.dp(6f)
        }
    }

    private fun bottomPadFor(row: MdRow): Float {
        val extra = if (row.isLastInMessage) theme.dp(20f) else 0f
        return extra + when (row.type) {
            RowType.HEADING -> theme.dp(6f)
            RowType.LIST_ITEM -> theme.dp(3f)
            else -> theme.dp(6f)
        }
    }

    // ---------------------------------------------------------------- bubble

    private fun buildBubble(row: MdRow, width: Int): RowLayout {
        val pad = theme.dp(13f)
        val maxBubble = width * 0.84f
        val maxText = max(1f, maxBubble - pad * 2)
        val spanned = spannable(row.inline, theme.bubble)
        val desired = measureMaxLineWidth(spanned, theme.bubble)
        val textW = max(1, min(desired, maxText).toInt())
        val layout = staticLayout(spanned, theme.bubble, textW, Layout.Alignment.ALIGN_NORMAL)
        val bubbleW = layout.let { l ->
            var w = 0f
            for (i in 0 until l.lineCount) w = max(w, l.getLineWidth(i))
            min(maxText, w) + pad * 2
        }
        val bubbleH = layout.height + pad * 2
        val top = theme.dp(if (row.isFirstInMessage) 14f else 4f)
        val bottom = theme.dp(10f)
        return BubbleRowLayout(
            text = spanned,
            layout = layout,
            bubbleLeft = width - hPad - bubbleW,
            bubbleTop = top,
            bubbleWidth = bubbleW,
            bubbleHeight = bubbleH,
            pad = pad,
            height = (top + bubbleH + bottom).toInt(),
        )
    }

    // ------------------------------------------------------------------ code

    private fun buildCode(row: MdRow, width: Int): RowLayout {
        val code = row.code ?: ""
        val pad = theme.dp(12f).toInt()
        val headerH = theme.dp(32f).toInt()
        val left = (hPad + row.quoteDepth * quoteInset).toInt()
        val boxWidth = width - left - hPad.toInt()

        val spanned = SpannableStringBuilder(code)
        for (t in CodeHighlighter.tokenize(code, row.language ?: "text")) {
            val color = when (t.kind) {
                CodeHighlighter.KIND_KEYWORD -> theme.synKeyword
                CodeHighlighter.KIND_STRING -> theme.synString
                CodeHighlighter.KIND_NUMBER -> theme.synNumber
                CodeHighlighter.KIND_COMMENT -> theme.synComment
                CodeHighlighter.KIND_TYPE -> theme.synType
                else -> theme.synFunc
            }
            if (t.end <= spanned.length && t.start < t.end) {
                spanned.setSpan(ForegroundColorSpan(color), t.start, t.end, Spanned.SPAN_EXCLUSIVE_EXCLUSIVE)
            }
        }

        // No wrapping: code scrolls horizontally, exactly like ChatGPT / Gemini.
        val natural = ceil(measureMaxLineWidth(spanned, theme.code)).toInt() + 2
        val contentWidth = max(natural, boxWidth - pad * 2)
        val layout = staticLayout(spanned, theme.code, contentWidth, Layout.Alignment.ALIGN_NORMAL)

        val vMargin = theme.dp(8f).toInt() + if (row.isFirstInMessage) theme.dp(10f).toInt() else 0
        val bottomMargin = theme.dp(8f).toInt() + if (row.isLastInMessage) theme.dp(20f).toInt() else 0
        val height = vMargin + headerH + pad + layout.height + pad + bottomMargin

        return CodeRowLayout(
            language = row.language ?: "text",
            code = code,
            text = spanned,
            layout = layout,
            headerHeight = headerH,
            contentWidth = contentWidth,
            padding = pad,
            left = left,
            boxWidth = boxWidth,
            topMargin = vMargin,
            height = height,
        )
    }

    // ----------------------------------------------------------------- table

    private fun buildTable(row: MdRow, width: Int): RowLayout {
        val table = row.table!!
        val cols = table.header.size
        val cellPad = theme.dp(10f).toInt()
        val left = (hPad + row.quoteDepth * quoteInset).toInt()
        val avail = width - left - hPad.toInt()
        val maxCol = theme.dp(230f)

        val allRows = ArrayList<List<InlineText>>(table.rows.size + 1)
        allRows.add(table.header)
        allRows.addAll(table.rows)

        val spanned = Array(allRows.size) { r ->
            Array(cols) { c ->
                val paint = if (r == 0) theme.tableHead else theme.tableCell
                spannable(allRows[r].getOrElse(c) { InlineText.EMPTY }, paint)
            }
        }

        val colW = IntArray(cols)
        for (c in 0 until cols) {
            var w = 0f
            for (r in spanned.indices) {
                val paint = if (r == 0) theme.tableHead else theme.tableCell
                w = max(w, measureMaxLineWidth(spanned[r][c], paint))
            }
            colW[c] = ceil(min(w, maxCol)).toInt() + cellPad * 2
        }

        // If the natural table is narrower than the viewport, stretch it to fill.
        val natural = colW.sum()
        if (natural < avail && natural > 0) {
            var used = 0
            for (c in 0 until cols) {
                val add = if (c == cols - 1) avail - natural - used else (avail - natural) * colW[c] / natural
                colW[c] += add
                used += add
            }
        }

        val cells = Array(spanned.size) { r ->
            Array(cols) { c ->
                val paint = if (r == 0) theme.tableHead else theme.tableCell
                val align = when (table.aligns.getOrElse(c) { ALIGN_LEFT }) {
                    ALIGN_CENTER -> Layout.Alignment.ALIGN_CENTER
                    ALIGN_RIGHT -> Layout.Alignment.ALIGN_OPPOSITE
                    else -> Layout.Alignment.ALIGN_NORMAL
                }
                staticLayout(spanned[r][c], paint, max(1, colW[c] - cellPad * 2), align)
            }
        }

        val rowH = IntArray(cells.size) { r ->
            var h = 0
            for (c in 0 until cols) h = max(h, cells[r][c].height)
            h + cellPad * 2
        }

        val vMargin = theme.dp(8f).toInt()
        val bottomMargin = theme.dp(8f).toInt() + if (row.isLastInMessage) theme.dp(20f).toInt() else 0
        return TableRowLayout(
            cells = cells,
            colW = colW,
            rowH = rowH,
            contentWidth = colW.sum(),
            cellPad = cellPad,
            left = left,
            topMargin = vMargin,
            height = vMargin + rowH.sum() + bottomMargin,
        )
    }

    // ----------------------------------------------------------------- utils

    fun spannable(inline: InlineText, base: TextPaint): Spanned {
        val sb = SpannableStringBuilder(inline.text)
        val n = sb.length
        for (s in inline.spans) {
            val start = s.start.coerceIn(0, n)
            val end = s.end.coerceIn(start, n)
            if (start == end) continue
            val f = Spanned.SPAN_EXCLUSIVE_EXCLUSIVE
            val bold = s.style and SPAN_BOLD != 0
            val italic = s.style and SPAN_ITALIC != 0
            if (bold && italic) sb.setSpan(StyleSpan(Typeface.BOLD_ITALIC), start, end, f)
            else if (bold) sb.setSpan(StyleSpan(Typeface.BOLD), start, end, f)
            else if (italic) sb.setSpan(StyleSpan(Typeface.ITALIC), start, end, f)
            if (s.style and SPAN_STRIKE != 0) sb.setSpan(StrikethroughSpan(), start, end, f)
            if (s.style and SPAN_CODE != 0) {
                sb.setSpan(TypefaceSpan("monospace"), start, end, f)
                sb.setSpan(RelativeSizeSpan(0.9f), start, end, f)
                sb.setSpan(BackgroundColorSpan(theme.inlineCodeBg), start, end, f)
                sb.setSpan(ForegroundColorSpan(theme.inlineCodeFg), start, end, f)
            }
            if (s.link != null) {
                sb.setSpan(ForegroundColorSpan(theme.link), start, end, f)
                sb.setSpan(MdLinkSpan(s.link), start, end, f)
            }
        }
        return sb
    }

    private fun measureMaxLineWidth(text: CharSequence, paint: TextPaint): Float {
        var maxW = 0f
        var start = 0
        val len = text.length
        while (start <= len) {
            var end = start
            while (end < len && text[end] != '\n') end++
            maxW = max(maxW, Layout.getDesiredWidth(text, start, end, paint))
            if (end >= len) break
            start = end + 1
        }
        return maxW
    }

    private fun staticLayout(
        text: CharSequence,
        paint: TextPaint,
        width: Int,
        align: Layout.Alignment,
    ): StaticLayout =
        StaticLayout.Builder.obtain(text, 0, text.length, paint, max(1, width))
            .setAlignment(align)
            .setLineSpacing(theme.dp(3f), 1.06f)
            .setIncludePad(false)
            // SIMPLE + no hyphenation is 3-6x faster than the default balanced
            // strategy and is what makes long answers cheap to lay out.
            .setBreakStrategy(Layout.BREAK_STRATEGY_SIMPLE)
            .setHyphenationFrequency(Layout.HYPHENATION_FREQUENCY_NONE)
            .build()
}

File: modules/react-native-md-list/ios/MdLayout.swift
import UIKit
import CoreText

extension NSAttributedString.Key {
    /// URL carried on a range; hit tested manually against the CTFrame.
    static let mdLink = NSAttributedString.Key("MdLink")
}

/// A laid out block of text.
///
/// CoreText rather than TextKit on purpose: CTFramesetter is thread safe, so the
/// expensive part (line breaking) happens on a worker queue and the main thread
/// only ever calls CTFrameDraw.
final class MdTextFrame {
    let attributed: NSAttributedString
    let ctFrame: CTFrame
    let size: CGSize

    init(_ attributed: NSAttributedString, width: CGFloat) {
        self.attributed = attributed
        let framesetter = CTFramesetterCreateWithAttributedString(attributed)
        var fitRange = CFRange()
        let constraints = CGSize(width: max(1, width), height: 100_000)
        let suggested = CTFramesetterSuggestFrameSizeWithConstraints(
            framesetter, CFRange(location: 0, length: 0), nil, constraints, &fitRange
        )
        let h = ceil(suggested.height)
        size = CGSize(width: ceil(suggested.width), height: h)
        let path = CGPath(rect: CGRect(x: 0, y: 0, width: max(1, width), height: max(1, h)), transform: nil)
        ctFrame = CTFramesetterCreateFrame(framesetter, CFRange(location: 0, length: 0), path, nil)
    }

    func draw(in ctx: CGContext, at origin: CGPoint) {
        ctx.saveGState()
        ctx.textMatrix = .identity
        ctx.translateBy(x: origin.x, y: origin.y + size.height)
        ctx.scaleBy(x: 1, y: -1)
        CTFrameDraw(ctFrame, ctx)
        ctx.restoreGState()
    }

    /// `point` is in UIKit coordinates relative to the frame's top-left corner.
    func link(at point: CGPoint) -> String? {
        guard let lines = CTFrameGetLines(ctFrame) as? [CTLine], !lines.isEmpty else { return nil }
        var origins = [CGPoint](repeating: .zero, count: lines.count)
        CTFrameGetLineOrigins(ctFrame, CFRange(location: 0, length: 0), &origins)
        let flipped = CGPoint(x: point.x, y: size.height - point.y)

        for (i, line) in lines.enumerated() {
            var ascent: CGFloat = 0, descent: CGFloat = 0, leading: CGFloat = 0
            let w = CGFloat(CTLineGetTypographicBounds(line, &ascent, &descent, &leading))
            let o = origins[i]
            let rect = CGRect(x: o.x, y: o.y - descent, width: w, height: ascent + descent)
            guard rect.contains(flipped) else { continue }
            let index = CTLineGetStringIndexForPosition(line, CGPoint(x: flipped.x - o.x, y: 0))
            guard index >= 0, index < attributed.length else { return nil }
            return attributed.attribute(.mdLink, at: index, effectiveRange: nil) as? String
        }
        return nil
    }
}

// MARK: - row layouts

class MdRowLayout {
    let height: CGFloat
    init(height: CGFloat) { self.height = height }
}

final class MdTextRowLayout: MdRowLayout {
    let frame: MdTextFrame
    let textOrigin: CGPoint
    let marker: MdTextFrame?
    let markerX: CGFloat
    let quoteDepth: Int
    let quoteX: CGFloat
    let checked: Int

    init(frame: MdTextFrame, textOrigin: CGPoint, marker: MdTextFrame?, markerX: CGFloat,
         quoteDepth: Int, quoteX: CGFloat, checked: Int, height: CGFloat) {
        self.frame = frame
        self.textOrigin = textOrigin
        self.marker = marker
        self.markerX = markerX
        self.quoteDepth = quoteDepth
        self.quoteX = quoteX
        self.checked = checked
        super.init(height: height)
    }
}

final class MdBubbleRowLayout: MdRowLayout {
    let frame: MdTextFrame
    let bubbleRect: CGRect
    let pad: CGFloat

    init(frame: MdTextFrame, bubbleRect: CGRect, pad: CGFloat, height: CGFloat) {
        self.frame = frame
        self.bubbleRect = bubbleRect
        self.pad = pad
        super.init(height: height)
    }
}

final class MdCodeRowLayout: MdRowLayout {
    let language: String
    let code: String
    let frame: MdTextFrame
    let headerHeight: CGFloat
    let contentWidth: CGFloat
    let padding: CGFloat
    let left: CGFloat
    let boxWidth: CGFloat
    let topMargin: CGFloat

    init(language: String, code: String, frame: MdTextFrame, headerHeight: CGFloat,
         contentWidth: CGFloat, padding: CGFloat, left: CGFloat, boxWidth: CGFloat,
         topMargin: CGFloat, height: CGFloat) {
        self.language = language
        self.code = code
        self.frame = frame
        self.headerHeight = headerHeight
        self.contentWidth = contentWidth
        self.padding = padding
        self.left = left
        self.boxWidth = boxWidth
        self.topMargin = topMargin
        super.init(height: height)
    }
}

final class MdTableRowLayout: MdRowLayout {
    let cells: [[MdTextFrame]]
    let colW: [CGFloat]
    let rowH: [CGFloat]
    let contentWidth: CGFloat
    let cellPad: CGFloat
    let left: CGFloat
    let topMargin: CGFloat

    init(cells: [[MdTextFrame]], colW: [CGFloat], rowH: [CGFloat], contentWidth: CGFloat,
         cellPad: CGFloat, left: CGFloat, topMargin: CGFloat, height: CGFloat) {
        self.cells = cells
        self.colW = colW
        self.rowH = rowH
        self.contentWidth = contentWidth
        self.cellPad = cellPad
        self.left = left
        self.topMargin = topMargin
        super.init(height: height)
    }
}

final class MdDividerRowLayout: MdRowLayout {
    let inset: CGFloat
    init(inset: CGFloat, height: CGFloat) {
        self.inset = inset
        super.init(height: height)
    }
}

// MARK: - engine

final class MdLayoutEngine {

    let theme: MdTheme
    let hPad: CGFloat = 16
    private let quoteInset: CGFloat = 14
    private let listIndent: CGFloat = 18
    private let markerGutter: CGFloat = 20

    private let cache = NSCache<NSString, MdRowLayout>()

    init(theme: MdTheme) {
        self.theme = theme
        cache.countLimit = 600
    }

    private func key(_ row: MdRow, _ width: CGFloat) -> NSString {
        "\(row.layoutKey)|\(Int(width))|\(theme.key)" as NSString
    }

    func cachedLayout(for row: MdRow, width: CGFloat) -> MdRowLayout? {
        cache.object(forKey: key(row, width))
    }

    @discardableResult
    func layout(for row: MdRow, width: CGFloat) -> MdRowLayout {
        let k = key(row, width)
        if let hit = cache.object(forKey: k) { return hit }
        let built = build(row, width: width)
        cache.setObject(built, forKey: k)
        return built
    }

    private func build(_ row: MdRow, width: CGFloat) -> MdRowLayout {
        switch row.type {
        case .bubble: return buildBubble(row, width: width)
        case .code: return buildCode(row, width: width)
        case .table: return buildTable(row, width: width)
        case .divider: return MdDividerRowLayout(inset: hPad, height: 25)
        default: return buildText(row, width: width)
        }
    }

    // MARK: text

    private func buildText(_ row: MdRow, width: CGFloat) -> MdRowLayout {
        let isQuote = row.quoteDepth > 0
        let font: UIFont = row.type == .heading ? theme.headingFont(row.headingLevel) : theme.body
        let color: UIColor = isQuote ? theme.quoteFg : theme.textPrimary

        var left = hPad + CGFloat(row.quoteDepth) * quoteInset
        var marker: MdTextFrame?
        var markerX: CGFloat = 0

        if row.type == .listItem {
            left += CGFloat(row.listDepth) * listIndent
            markerX = left
            if row.checked < 0, let m = row.listMarker {
                marker = MdTextFrame(
                    attributed(MdInlineText(text: m, spans: []), font: font, color: theme.textSecondary),
                    width: markerGutter
                )
            }
            left += markerGutter
        }

        let avail = max(1, width - left - hPad)
        let string = attributed(row.inline, font: font, color: color)
        let frame = MdTextFrame(string, width: avail)

        let topPad = topPadding(row)
        let bottomPad = bottomPadding(row)

        return MdTextRowLayout(
            frame: frame,
            textOrigin: CGPoint(x: left, y: topPad),
            marker: marker,
            markerX: markerX,
            quoteDepth: row.quoteDepth,
            quoteX: hPad,
            checked: row.checked,
            height: ceil(topPad + frame.size.height + bottomPad)
        )
    }

    private func topPadding(_ row: MdRow) -> CGFloat {
        let extra: CGFloat = row.isFirstInMessage ? 10 : 0
        switch row.type {
        case .heading:
            switch row.headingLevel {
            case 1: return extra + 18
            case 2: return extra + 16
            case 3: return extra + 14
            default: return extra + 12
            }
        case .listItem: return extra + 3
        default: return extra + 6
        }
    }

    private func bottomPadding(_ row: MdRow) -> CGFloat {
        let extra: CGFloat = row.isLastInMessage ? 20 : 0
        switch row.type {
        case .heading: return extra + 6
        case .listItem: return extra + 3
        default: return extra + 6
        }
    }

    // MARK: bubble

    private func buildBubble(_ row: MdRow, width: CGFloat) -> MdRowLayout {
        let pad: CGFloat = 13
        let maxText = max(1, width * 0.84 - pad * 2)
        let string = attributed(row.inline, font: theme.body, color: theme.textPrimary)
        let frame = MdTextFrame(string, width: maxText)
        let bubbleW = min(maxText, frame.size.width) + pad * 2
        let bubbleH = frame.size.height + pad * 2
        let top: CGFloat = row.isFirstInMessage ? 14 : 4
        let bottom: CGFloat = 10
        return MdBubbleRowLayout(
            frame: frame,
            bubbleRect: CGRect(x: width - hPad - bubbleW, y: top, width: bubbleW, height: bubbleH),
            pad: pad,
            height: ceil(top + bubbleH + bottom)
        )
    }

    // MARK: code

    private func buildCode(_ row: MdRow, width: CGFloat) -> MdRowLayout {
        let code = row.code ?? ""
        let pad: CGFloat = 12
        let headerH: CGFloat = 32
        let left = hPad + CGFloat(row.quoteDepth) * quoteInset
        let boxWidth = width - left - hPad

        let string = NSMutableAttributedString(string: code)
        let para = NSMutableParagraphStyle()
        para.lineSpacing = 2
        para.lineBreakMode = .byClipping
        string.addAttributes(
            [.font: theme.codeFont, .foregroundColor: theme.codeFg, .paragraphStyle: para],
            range: NSRange(location: 0, length: string.length)
        )
        for token in MdCodeHighlighter.tokenize(code, language: row.language ?? "text") {
            let location = max(0, min(token.start, string.length))
            let length = max(0, min(token.end, string.length) - location)
            guard length > 0 else { continue }
            string.addAttribute(
                .foregroundColor,
                value: color(for: token.kind),
                range: NSRange(location: location, length: length)
            )
        }

        // measure unwrapped, then let the row scroll horizontally
        let natural = MdTextFrame(string, width: 100_000).size.width
        let contentWidth = max(natural, boxWidth - pad * 2)
        let frame = MdTextFrame(string, width: contentWidth)

        let topMargin: CGFloat = 8 + (row.isFirstInMessage ? 10 : 0)
        let bottomMargin: CGFloat = 8 + (row.isLastInMessage ? 20 : 0)

        return MdCodeRowLayout(
            language: row.language ?? "text",
            code: code,
            frame: frame,
            headerHeight: headerH,
            contentWidth: contentWidth,
            padding: pad,
            left: left,
            boxWidth: boxWidth,
            topMargin: topMargin,
            height: ceil(topMargin + headerH + pad * 2 + frame.size.height + bottomMargin)
        )
    }

    private func color(for kind: MdCodeHighlighter.Kind) -> UIColor {
        switch kind {
        case .keyword: return theme.synKeyword
        case .string: return theme.synString
        case .number: return theme.synNumber
        case .comment: return theme.synComment
        case .type: return theme.synType
        case .function: return theme.synFunc
        }
    }

    // MARK: table

    private func buildTable(_ row: MdRow, width: CGFloat) -> MdRowLayout {
        guard let table = row.table else { return MdDividerRowLayout(inset: hPad, height: 0) }
        let cols = table.header.count
        let cellPad: CGFloat = 10
        let left = hPad + CGFloat(row.quoteDepth) * quoteInset
        let avail = width - left - hPad
        let maxCol: CGFloat = 230

        var allRows: [[MdInlineText]] = [table.header]
        allRows.append(contentsOf: table.rows)

        let strings: [[NSAttributedString]] = allRows.enumerated().map { (r, cells) in
            (0..<cols).map { c in
                let font = r == 0 ? theme.tableHeadFont : theme.tableCellFont
                let alignment: NSTextAlignment
                switch table.aligns.indices.contains(c) ? table.aligns[c] : .left {
                case .center: alignment = .center
                case .right: alignment = .right
                case .left: alignment = .natural
                }
                return attributed(
                    c < cells.count ? cells[c] : .empty,
                    font: font, color: theme.textPrimary, alignment: alignment
                )
            }
        }

        var colW = [CGFloat](repeating: 0, count: cols)
        for c in 0..<cols {
            var w: CGFloat = 0
            for r in strings.indices {
                w = max(w, MdTextFrame(strings[r][c], width: 100_000).size.width)
            }
            colW[c] = ceil(min(w, maxCol)) + cellPad * 2
        }

        let natural = colW.reduce(0, +)
        if natural < avail, natural > 0 {
            var used: CGFloat = 0
            for c in 0..<cols {
                let add = c == cols - 1 ? (avail - natural - used) : (avail - natural) * colW[c] / natural
                colW[c] += add
                used += add
            }
        }

        let cells: [[MdTextFrame]] = strings.map { rowStrings in
            (0..<cols).map { c in MdTextFrame(rowStrings[c], width: max(1, colW[c] - cellPad * 2)) }
        }
        let rowH: [CGFloat] = cells.map { rowFrames in
            (rowFrames.map { $0.size.height }.max() ?? 0) + cellPad * 2
        }

        let topMargin: CGFloat = 8
        let bottomMargin: CGFloat = 8 + (row.isLastInMessage ? 20 : 0)
        return MdTableRowLayout(
            cells: cells,
            colW: colW,
            rowH: rowH,
            contentWidth: colW.reduce(0, +),
            cellPad: cellPad,
            left: left,
            topMargin: topMargin,
            height: ceil(topMargin + rowH.reduce(0, +) + bottomMargin)
        )
    }

    // MARK: attributed string

    func attributed(
        _ inline: MdInlineText,
        font: UIFont,
        color: UIColor,
        alignment: NSTextAlignment = .natural
    ) -> NSAttributedString {
        let string = NSMutableAttributedString(string: inline.text)
        let full = NSRange(location: 0, length: string.length)
        guard full.length > 0 else { return string }

        let para = NSMutableParagraphStyle()
        para.lineSpacing = 3
        para.alignment = alignment
        para.lineBreakMode = .byWordWrapping
        string.addAttributes([.font: font, .foregroundColor: color, .paragraphStyle: para], range: full)

        for span in inline.spans {
            let location = max(0, min(span.start, string.length))
            let length = max(0, min(span.end, string.length) - location)
            guard length > 0 else { continue }
            let range = NSRange(location: location, length: length)

            if span.style.contains(.code) {
                string.addAttributes([
                    .font: UIFont.monospacedSystemFont(ofSize: font.pointSize * 0.9, weight: .regular),
                    .foregroundColor: theme.inlineCodeFg,
                    .backgroundColor: theme.inlineCodeBg,
                ], range: range)
            } else if span.style.contains(.bold) || span.style.contains(.italic) {
                var updates: [(NSRange, UIFont)] = []
                string.enumerateAttribute(.font, in: range, options: []) { value, sub, _ in
                    let current = (value as? UIFont) ?? font
                    updates.append((sub, withTraits(
                        current,
                        bold: span.style.contains(.bold),
                        italic: span.style.contains(.italic)
                    )))
                }
                for (sub, f) in updates { string.addAttribute(.font, value: f, range: sub) }
            }

            if span.style.contains(.strike) {
                string.addAttribute(.strikethroughStyle, value: NSUnderlineStyle.single.rawValue, range: range)
            }
            if let link = span.link {
                string.addAttributes([.foregroundColor: theme.link, .mdLink: link], range: range)
            }
        }
        return string
    }

    private func withTraits(_ font: UIFont, bold: Bool, italic: Bool) -> UIFont {
        var traits = font.fontDescriptor.symbolicTraits
        if bold { traits.insert(.traitBold) }
        if italic { traits.insert(.traitItalic) }
        guard let descriptor = font.fontDescriptor.withSymbolicTraits(traits) else { return font }
        return UIFont(descriptor: descriptor, size: font.pointSize)
    }
}
